### PR TITLE
 add mask to loss

### PR DESCRIPTION
**Pull request by marouaneamz**
this pull request adds option of the mask to the loss which allows not to take  into account  images in the loss if they do not have labels for certain task

### CODE REVIEW

File: mmcls/datasets/pipelines/formatting.py [hunk 1]
```diff
@@ -65,6 +65,46 @@ def __repr__(self):
...
+                    gt_mask[task] = v
+        else:
+            for task in self.tasks:
+                label_key = task + '_img_label'
+                mask_key = task + '_mask'
+                gt_label[task] = to_tensor(results[label_key])
+                gt_mask[task] = results[mask_key]
+
+
+        results['gt_label'] = dict(label=gt_label,mask=gt_mask)
```
Comment (piercus) [inline mmcls/datasets/pipelines/formatting.py]:
```suggestion
                gt_label[task] = dict(label=to_tensor(results[label_key]), mask=results[mask_key])


        results['gt_label'] = gt_label
```

File: mmcls/models/heads/multi_task_head.py [hunk 2]
```diff
@@ -31,10 +31,14 @@ def __init__(self, sub_heads, common_cfg=dict(), init_cfg=None):
             sub_head = HEADS.build(head_cfg, default_args=common_cfg)
             self.sub_heads[task_name] = sub_head
 
-    def forward_train(self, x, gt_label, **kwargs):
+    def forward_train(self, features, gt_label, **kwargs):
         losses = dict()
         for task_name, head in self.sub_heads.items():
-            head_loss = head.forward_train(x, gt_label[task_name], **kwargs)
+            mask = gt_label['mask'][task_name]
```
Comment (piercus) [inline mmcls/models/heads/multi_task_head.py]:
```suggestion
            task_gt_label = gt_label[task_name]
            if type(task_gt_label) is dict :
              mask = task_gt_label['mask']
              label = task_gt_label['label']
            else: # a tensor
              label = task_gt_label['label']
              batch_n = label.shape[0]
              mask = to_tensor([True]*batch_n)
```

**Review by piercus — CHANGES_REQUESTED** [review]
@marouaneamz please see inline comments

**Comment by marouaneamz** [comment]
I try to adapt the pipeline if there is no mask , but I can't do your method because we add the labels with the first if (if we haven't hardcoded the tasks [here](https://github.com/piercus/mmclassification/blob/multi-task/mmcls/datasets/pipelines/formatting.py#L52) ), and in this block we add the labels and the maks we go through the whole results list and not all of them at once

File: mmcls/datasets/pipelines/formatting.py [hunk 3]
```diff
@@ -65,6 +65,46 @@ def __repr__(self):
...
+        gt_label = dict()
+        gt_mask = dict()
+        if self.tasks is None:
+            for k, v in results.items():
+                if k.endswith('_img_label'):
+                    task = k[:-10]
+                    gt_label[task] = to_tensor(v)
+                elif k.endswith('_mask'):
+                    task = k[:-5]
+                    gt_mask[task] = v
```
Comment (piercus) [inline mmcls/datasets/pipelines/formatting.py]:
```suggestion
            for k, v in results.items():
                if k.endswith('_img_label'):
                    task = k[:-10]
                    if task not in gt_label:
                      gt_label[task] = dict()
                    gt_label[task]['label'] = to_tensor(v)
                elif k.endswith('_mask'):
                    task = k[:-5]
                    if task not in gt_label:
                      gt_label[task] = dict()
                    gt_label[task]['mask'] = v
```

**Review by piercus — APPROVED** [review]
